### ISSUE TITLE
[discussion] Abstract messaging and state modifications

### ISSUE THREAD

**Issue by Andarist** [ivantsov/redux-webext#5]
First of all - great lib!

I was wondering about such solution myself, but for a little bit different use case. My target are iframes/workers and not web extensions. Messaging API is really quite similar, so I think that could be quite easily abstracted to serve more general use case.

For now I'm using a custom postMessage middleware which passes actions from one context to another, but I keep regular stores in both contexts, so they just exchange actions but both recalculate the state (sometimes differently, but in general I try to keep them in synced, but the one is kinda a subset of the another).

Big concern of my with ur solution is that it exchanges the whole state each time which is not that performant, especially for bigger states - copying them from the context to context is not even that much of a concern probably, should be quite ok. But recreating it on the other side (consumer's) destroys all references so the `react-redux` wont be able to act too performantly.

I was thinking about the solution for this. If every reducer could return a tuple of `[state, modification]` (which could be abstracted with a helpers to update state - like set, concat, push, etc + higher order reducer which would add meta to those modifications per state slice). Finally ofc reducer would return only the updated state, but collecting modifications along the way would allow us to send them to consumers which could just interpret them and update the state in a more performant fashion.

WDYT?

**Comment by ivantsov** [comment]
Hi @Andarist,

Thank you for your interest in the project and feedback!

> My target are iframes/workers and not web extensions. Messaging API is really quite similar, so I think that could be quite easily abstracted to serve more general use case.

Honestly, I'm not that much aware about the `messaging` API but if you think it's quite similar to `WebExtension` API - then I guess we can add it to the lib.

> For now I'm using a custom postMessage middleware which passes actions from one context to another, but I keep regular stores in both contexts.

If I understand correctly you have several real redux stores in different parts of the app. However, it brings 2 sources of truth, so as you mentioned it leads to the problem of their synchronization. This's why the lib encourages you to have only one real store in _background_ page and _ui_ stores can consume this data but don't influence on it in any way.

> Big concern of my with ur solution is that it exchanges the whole state each time which is not that performant, especially for bigger states.

I haven't faced any problems with exchanging the whole state between different parts of the app so far. But to be honest, I don't use a really big state (it's hard to say what everybody means by "big state"). Anyway, I wouldn't start raising concerns about performance until it really becomes a problem. At least it's hard to reason about solving the problem without any specific case.

> I was thinking about the solution for this.
That might be a solution for enhancing the performance however as I said above, I don't think we should address the problem which doesn't really exist at the moment.

To recap, I think we should keep the lib as simple as possible and try to preserve simple API without introducing something new like _helpers to update state_. The initial goal of the lib was and remains to have an ability to use, familiar to everybody, redux for WebExtensions without learning something new.
Nevertheless, I'm happy to discuss exposing lib on the `messaging` API.

**Comment by Andarist** [comment]
So what I have meant with Messaging API was in general any postMessage-like API. There are at least 2 other similar ones to the one implemented by WebExt:
- https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API
- https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage

They are only slighly different. We could abstract the core API - `postMessage` + `onmessage`, so the core of the lib would accept such interface and make adapters for those 3 basic ones. That would also allow connecting any other messaging interface by just adapting it to what is acceptable by the core part.

About performance - yeah, thats truth thats its hard to talk about it without data, which unfortunately i dont have. However Im struggling with the perf from time to time in regular redux architecture - so single store, references intact etc, especially when dealing with dynamic lists updated in real time - think of chat list, feed list, markers on the map.

Its anyway certainly not the time for me to refactor the project to use ur lib or the proposed approach - just brainstorming for now ;)

**Comment by ivantsov** [comment]
Sorry for not answering too long, summer you know :)

I can suggest two options:
- when I have time I'll start discovering messages API and add to the lib, however I can't guarantee any time frames
- you can prepare a PR

**Comment by Andarist** [comment]
I might look into that, but also cannot make any promises as I have work to do in some other projects first :)
